Add loading mask to d2-ui-components

SferaDev opened the pull request:
### :memo: Implementation

- Add a Loading Mask Provider

### :art: Screenshots

![image](https://user-images.githubusercontent.com/2181866/54998915-8b92f780-4fcf-11e9-9c04-d421b14cbbbb.png)

SferaDev committed:
Add loading mask to d2-ui-components

tokland commented on README.md:
```diff
@@ -121,6 +121,44 @@ MyComponent.propTypes = {
+const MyComponent = ({name, loading}) => (
+    <div>
+        <a onClick={() => loading.show()}>Show loading mask</a>
+        <a onClick={() => loading.show(false)}>Hide loading mask</a>
```
`loading.show(boolean)` is very useful, but can we have also `loading.hide()` for the simple case?

SferaDev replied:
Sounds good, will add.

tokland commented on src/loading/LoadingConsumer.js:
```diff
@@ -0,0 +1,81 @@
+const LoadingConsumer = props => {
+    const { classes } = props;
+
+    const showMessage = () => {
```
Unused function.

tokland commented on src/loading/LoadingConsumer.js:
```diff
@@ -0,0 +1,81 @@
+                            <Typography
+                                className={classes.message}
+                                variant="h6"
+                                hidden={!message || message.trim().length === 0}
```
Used two times, maybe move it a to local variable `hidden`?

SferaDev replied:
Okay I'll move to render variable

tokland replied:
Now that we are at it, isn't equivalent to: `!message || !message.trim()`?

tokland commented on src/loading/LoadingConsumer.js:
```diff
@@ -0,0 +1,81 @@
+import React from "react";
```
Should be a jsx file. I see why you have a js file, I had it wrong for the snack: `src/snackbar/SnackbarConsumer.js`, could be changed also there.

SferaDev replied:
Yeah I just started the branch off Snackbar with shallow copy. Will change them in this PR now

tokland commented on src/loading/LoadingConsumer.js:
```diff
@@ -0,0 +1,81 @@
+import Typography from "@material-ui/core/Typography";
+import Divider from "@material-ui/core/Divider";
+
+const styles = theme => ({
```
eslint does not like non-used arguments, `_theme` should do the trick.

tokland commented on src/loading/LoadingProvider.js:
```diff
@@ -0,0 +1,55 @@
+import React, { Component } from "react";
```
Same, jsx file.

tokland commented on src/loading/LoadingProvider.js:
```diff
@@ -0,0 +1,55 @@
+import LoadingContext from "./context";
+import LoadingConsumer from "./LoadingConsumer";
+
+export default class SnackbarProvider extends Component {
```
LoadingProvider

tokland reviewed (commented):
The code (and the snapshot, nice!) looks good to me, just some minor things.

Btw, Is there a branch in metadata-sync that uses it, so we can check how it works? Sometime we should create a simple dirty app with all the components on one page so we can them easily.

SferaDev committed:
Minor changes

SferaDev committed:
Set version to classnames

SferaDev commented:
> Btw, Is there a branch in metadata-sync that uses it, so we can check how it works? Sometime we should create a simple dirty app with all the components on one page so we can them easily.

Not yet but we should have it before Friday. Anyway I think the best approach is creating a [StoryBook](https://github.com/storybooks/storybook). It's really cool and now it's like the state-of-the-art solution for ReactJS examples.

tokland commented:
> Anyway I think the best approach is creating a [StoryBook](https://github.com/storybooks/storybook). 

Wow, it really looks nice. I think it still makes sense to  have an `example` CRA in the repo itself, but it looks good. Maybe you can show us on the next team meeting.

tokland reviewed (approved):
I have not tested functionally, I'll wait for CRA example (or metadata-sync test branch), but the code looks very good to me.

SferaDev committed:
Prettify code

SferaDev committed:
Add ConfirmationDialog and update DialogButton

SferaDev committed:
Allow to send props to mui-dialog

SferaDev committed:
Move children to DialogContent and prettify

SferaDev committed:
Simplify onClose

SferaDev committed:
MultiSelector: Make some props optional

SferaDev committed:
MultiSelector: Allow selection to be uncontrolled

SferaDev committed:
LoadingConsumer: Increase zIndex

Set the zIndex to be on top of all the arbitrary indexes of material-ui.

This fixes the loading to be below modal dialogs.

tokland committed:
Add argsIgnorePattern _ for no-unused-vars eslint rule

tokland committed:
Add missing onChange propType

tokland committed:
Refactor MultiSelector controlled/uncontrolled mode

SferaDev committed:
Simplify controlled/uncontrolled

SferaDev committed:
Re-add promise resolve

SferaDev committed:
Do not update state when not controlled

SferaDev committed:
Improve dialog component

SferaDev committed:
Add readme entry for ConfirmationDialog

adrianq committed:
Merge pull request #22 from EyeSeeTea/feature/confirmation-dialog

Abstract Confirmation dialog and update Dialog Button

adrianq committed:
Merge pull request #23 from EyeSeeTea/feature/multi-selector

MultiSelector: Allow selection to be uncontrolled

adrianq reviewed (approved):
I tested all the features in the multiple metadata-sync PR. Awesome work guys!
